kuma/scrape/scraper.py

```diff
@@ -0,0 +1,174 @@
+
+    source_types = {
+        'user': UserSource,
+    }
+
+    def __init__(self, host='developer.mozilla.org', ssl=True):
+        """Initialize Scraper."""
+        self.requester = Requester(host, ssl)
+        self.sources = OrderedDict()
+        self.storage = Storage()
+        self.defaults = {}
+        self.overrides = {}
+
+    def add_source(self, source_type, source_param='', **options):
+        """Add a source of MDN data."""
+        source_key = "%s:%s" % (source_type, source_param)
+        if source_key in self.sources:
+            changes = self.sources[source_key].merge_options(**options)
+            if changes:
+                logger.debug('Updating source "%s" with options %s',
+                             source_key, changes)
+                return True
+            else:
+                return False
+        else:
+            logger.debug('Adding source "%s" with options %s', source_key, options)
+            source_options = self.defaults.get(source_type, {})
+            source_options.update(options)
+            source = self.create_source(source_key, **source_options)
+            self.sources[source_key] = source
+            return True
+
+    def create_source(self, source_key, **options):
+        source_type, source_param = source_key.split(':', 1)
+        return self.source_types[source_type](source_param, **options)
+
+    def scrape(self):
+        """Scrape data from MDN sources."""
+        if not self.sources:
+            logger.warn("No sources to scrape.")
+            return self.sources
+        first = True
+        repeat = False
+        cycle = 0
+        state_counts = OrderedDict((state, 0) for state in Source.STATES)
+        state_counts[Source.STATE_INIT] = len(self.sources)
+        while first or repeat:
+            first = False
+            repeat = False
+            source_total = (len(self.sources) -
+                            state_counts[Source.STATE_DONE] -
+                            state_counts[Source.STATE_ERROR])
+            last_counts = state_counts
+            state_counts = OrderedDict((state, 0) for state in Source.STATES)
+            new_sources = []
+            cycle += 1
+
+            # Iterate over existing sources
+            source_num = 0
+            for source_key, source in self.sources.items():
+
+                # If terminal condition, no processing to do
+                if source.state in (Source.STATE_DONE, Source.STATE_ERROR):
+                    state_counts[source.state] += 1
+                    continue
+
+                # Gather dependent sources
+                source_num += 1
+                old_state = source.state
+                dependencies = source.gather(self.requester, self.storage)
+                new_sources.extend(dependencies)
+                dep_count = len(dependencies)
+                state_counts[source.state] += 1
+                for dep in dependencies:
+                    if "%" in dep[1]:
```

Comment:
It's not clear to me why this is needed, but perhaps it'll be more clear in the follow-on PR's?